packages/js/src/utils/generators.ts

```diff
@@ -0,0 +1,36 @@
+/** @link [Proof Key for Code Exchange by OAuth Public Clients](https://datatracker.ietf.org/doc/html/rfc7636) */
+
+import { fromUint8Array } from 'js-base64';
+
+const DEFAULT_RANDOM_DATA_LENGTH = 64;
+
+/**
+ * @param length The length of the raw random data.
+ */
+function generateRandomString(length: number) {
+  return fromUint8Array(crypto.getRandomValues(new Uint8Array(length)), true);
+}
+
+/**
+ * Generates random string for state and encodes them in url safe base64
+ */
+export const generateState = () => generateRandomString(DEFAULT_RANDOM_DATA_LENGTH);
+
+/**
+ * Generates code verifier
+ *
+ * @link [Client Creates a Code Verifier](https://datatracker.ietf.org/doc/html/rfc7636#section-4.1)
+ */
+export const generateCodeVerifier = () => generateRandomString(DEFAULT_RANDOM_DATA_LENGTH);
+
+/**
+ * Calculates the S256 PKCE code challenge for an arbitrary code verifier and encodes it in url safe base64
+ *
+ * @param {String} codeVerifier Code verifier to calculate the S256 code challenge for
+ * @link [Client Creates the Code Challenge](https://datatracker.ietf.org/doc/html/rfc7636#section-4.2)
+ */
+export const generateCodeChallenge = async (codeVerifier: string): Promise<string> => {
+  const encodedCodeVerifier = new TextEncoder().encode(codeVerifier);
+  const codeChallenge = new Uint8Array(await crypto.subtle.digest('SHA-256', encodedCodeVerifier));
+  return fromUint8Array(codeChallenge, true);
```

Comment:
we can remove related func in `essentials` if needed

Reply:
no hurries too

Reply:
ditto: use UrlSafeBase64 in #128 , you can take a look